common/telemetry/grpc.go

```diff
@@ -31,84 +31,151 @@ import (
+// NewServerStatsHandler creates a new gRPC server interceptor that tracks
+// each request with an encapsulating span using the provided TracerProvider and
+// TextMapPropagator.
+//
+// NOTE: If the TracerProvider is not enabled (i.e. is a noop provider), it returns `nil`.
+func NewServerStatsHandler(
+	tp trace.TracerProvider,
+	tmp propagation.TextMapPropagator,
+	logger log.Logger,
+) ServerStatsHandler {
+	if !isEnabled(tp) {
+		return nil
+	}
+
+	return newCustomServerStatsHandler(
+		otelgrpc.NewServerHandler(
+			otelgrpc.WithPropagators(tmp),
+			otelgrpc.WithTracerProvider(tp),
+		),
+		logger)
+}
+
+// NewClientStatsHandler creates a new gRPC client interceptor that tracks
 // each request with an encapsulating span using the provided TracerProvider and
 // TextMapPropagator.
-func NewServerTraceInterceptor(
+//
+// NOTE: If the TracerProvider is not enabled (i.e. is a noop provider), it returns `nil`.
+func NewClientStatsHandler(
 	tp trace.TracerProvider,
 	tmp propagation.TextMapPropagator,
-) ServerTraceInterceptor {
-	//nolint:staticcheck
-	otelInterceptor := otelgrpc.UnaryServerInterceptor(
+) ClientStatsHandler {
+	if !isEnabled(tp) {
+		return nil
+	}
+
+	return otelgrpc.NewClientHandler(
 		otelgrpc.WithPropagators(tmp),
 		otelgrpc.WithTracerProvider(tp),
 	)
+}
 
-	if debugMode() {
-		return func(
-			ctx context.Context,
-			req interface{},
-			info *grpc.UnaryServerInfo,
-			handler grpc.UnaryHandler,
-		) (any, error) {
-			return otelInterceptor(ctx, req, info, debugHandler(handler))
-		}
+func isEnabled(tp trace.TracerProvider) bool {
+	tracer := tp.Tracer("check-tracer")
+	_, span := tracer.Start(context.Background(), "check-span")
+	defer span.End()
+	return span.IsRecording()
+}
+
+func newCustomServerStatsHandler(
+	handler stats.Handler,
+	logger log.Logger,
+) *customServerStatsHandler {
+	return &customServerStatsHandler{
+		wrapped: handler,
+		isDebug: debugMode(),
+		tags:    logtags.NewWorkflowTags(common.NewProtoTaskTokenSerializer(), logger),
 	}
+}
 
-	return ServerTraceInterceptor(otelInterceptor)
+func (c *customServerStatsHandler) TagRPC(ctx context.Context, info *stats.RPCTagInfo) context.Context {
+	return c.wrapped.TagRPC(
+		context.WithValue(ctx, methodNameKey, info.FullMethodName),
+		info)
 }
 
-func debugHandler(origHandler grpc.UnaryHandler) grpc.UnaryHandler {
-	return func(ctx context.Context, req any) (resp any, err error) {
-		resp, err = origHandler(ctx, req)
+func (c *customServerStatsHandler) HandleRPC(ctx context.Context, stat stats.RPCStats) {
+	c.wrapped.HandleRPC(ctx, stat)
+
+	switch s := stat.(type) {
+	case *stats.InPayload:
 		span := trace.SpanFromContext(ctx)
-		if span.IsRecording() {
+
+		methodName, ok := ctx.Value(methodNameKey).(string)
+		if !ok {
+			methodName = "unknown"
+		}
+
+		// annotate span with workflow tags (same ones the Temporal SDKs use)
+		for _, tag := range c.tags.Extract(s.Payload, methodName) {
```

Comment:
This is re-using the existing generated code that can extract Workflow tags from gRPC requests. It seems to do exactly what we need, so I'm not inclined to clone/fork that code.